axiom/upgrade.py

```diff
@@ -253,6 +254,80 @@ def upgradeAllTheWay(o):
     return o
 
 
+
+def _hasExplicitOid(store, table):
+    """
+    Does the given table have an explicit oid column?
+    """
+    return any(info[1] == 'oid' for info
+               in store.querySchemaSQL(
+                   'PRAGMA *DATABASE*.table_info({})'.format(table)))
+
+
+
+def _upgradeTableOid(store, table, createTable, postCreate=lambda: None):
+    """
+    Upgrade a table to have an explicit oid.
+
+    Must be called in a transaction to avoid corrupting the database.
+    """
+    if _hasExplicitOid(store, table):
+        return
+    store.executeSchemaSQL(
+        'ALTER TABLE *DATABASE*.{0} RENAME TO {0}_temp'.format(table))
+    createTable()
+    store.executeSchemaSQL(
+        'INSERT INTO *DATABASE*.{0} '
+        'SELECT oid, * FROM *DATABASE*.{0}_temp'.format(table))
+    store.executeSchemaSQL('DROP TABLE *DATABASE*.{0}_temp'.format(table))
+    postCreate()
+
+
+
+def upgradeSystemOid(store):
```

Comment:
This may be outside the scope of this PR, but given that this upgrader copies an entire table, do you have any concerns about running out of storage space?  I suppose this problem is the same for all upgraders, since there is no such thing as an "in-place" upgrader in Axiom.

Also, would it make sense to VACUUM the entire database after this upgrade is complete?  Seems to me that would be a nice thing to do, or to at least provide the user the option to do so when upgrading.  Maybe it is sufficient to just emit a message when the upgrade is complete, suggesting that the user perform the VACUUM manually?

Reply:
I plan to tackle VACUUMing concerns in #84. I'd also like to make item deletion really delete items from `axiom_objects` but since this is only safe to do on an upgraded store, I figure that belongs in another PR too.